windows-agent/cmd/ubuntu-pro-agent/agent/agent_test.go

```diff
@@ -23,6 +24,64 @@ func init() {
 	daemontestutils.DefaultNetworkDetectionToMock()
 }
 
+func TestSingleInstance(t *testing.T) {
+	t.Parallel()
+
+	testcases := map[string]struct {
+		args                   []string
+		anotherInstanceRunning bool
+		breakLockFile          bool
+
+		wantError bool
+	}{
+		"Success when single instance": {args: []string{"version"}},
+
+		// Testing the 'clean' verb cannot run in parallel because it tries to delete files still opened by other test cases.
```

Comment:
We should definitively have a test for clean, though, even if it’s a separate test case, no? I’m unsure to understand as we have separate private directory for the lock file why we can’t run them in parallel.

Reply:
The 'clean' verb removes agent data from both public and private directories, but it has no knowledge of the public and private dir options, as the options are not cached inside the App struct. Thus, it goes through `os.Getenv()` to figure out those directories and end up mesing with global state affecting other tests (not only in this test function btw). Some refactoring is needed to lift computation of the options so they can be passed to `a.installClean()` to make sure that verb respects the temporary directories created by the test cases as well as changing the `TestClean` function to take into account that we should no longer solely rely on the environment variables.

So, for now, I'm taking the easy approach of adding one test case in `TestClean`  (which is already forbidden from running in parallel) that checks whether the agent proceeds with 'clean' or bails out if a lock file exists.

Reply:
Makes sense, thanks for detailing it!